### PR TITLE
Add unit tests for multiple python versions to CI stage

### PR DESCRIPTION
**Pull request by fpgmaas**
This PR builds upon [this PR](https://github.com/MrPowers/chispa/pull/80). When that is merged I can update this branch to better reflect the changes in "Files changed".

Related issue: https://github.com/MrPowers/chispa/issues/82

---

This PR adds testing against multiple Python versions, updates some of the mkdocs dependencies, and changes the trigger of the CI checks to run on each pull request and on each commit to master (i.e. merging a PR). For an example run, see [here](https://github.com/fpgmaas/chispa/actions/runs/6564589223), or approve the workflow run on this PR.

This also adds a check for the `poetry.lock` file; to prevent issues like #107 

I think this can be helpful, let me know what you think of the proposal!

### CODE REVIEW

**Comment by MrPowers** [comment]
@fpgmaas - really sorry for never responding here.  Should we have merged this one?

**Comment by fpgmaas** [comment]
@MrPowers  No worries; I think that is my own doing, since you pointed out in another PR that you did not want to drop support for Python 3.7 just yet, and then life got in the way and I didn't get around to updating it. I'll take a look at it again somewhere this week and try to modify & reopen PR's that could still be useful.

**Comment by MrPowers** [comment]
@fpgmaas - yea, thanks for understanding.

We're actually dropping support for PySpark 2 on quinn, so now is probably a good time to make these changes for chispa as well.  Let's brainstorm the next steps for this project!

**Comment by fpgmaas** [comment]
@MrPowers I updated [the PR](https://github.com/MrPowers/chispa/pull/80) that this one is built on top of and rebased this one. Example run can be found [here](https://github.com/fpgmaas/chispa/actions/runs/9930074399).

**Comment by fpgmaas** [comment]
@SemyonSinchenko just a heads up that this is an old PR so it won't show up at the top of open PR's. If you have time, could you review? No hurry! :)

**Review by SemyonSinchenko — COMMENTED** [review]
I left few comments

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,39 +1,56 @@
 name: Unit tests
 
 on:
+  pull_request:
```
[the remark this answers is not in the document]
Comment (fpgmaas) [reply .github/workflows/ci.yml:4]: @SemyonSinchenko I believe that limit only holds for private repositories; public repositories have unlimited build minutes.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,39 +1,56 @@
...
+
+      - name: Check lock file
+        run: poetry lock --check
 
   test:
     runs-on: ubuntu-latest
-
+    strategy:
+      matrix:
+        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
```
[the remark this answers is not in the document]
Comment (fpgmaas) [reply .github/workflows/ci.yml:26]: We should definitely test for different Python versions to ensure that `chispa` works for each Python version. In addition we can also test for multiple versions of Pyspark, as I see you did in `quinn`. 

Ideally, we test for multipel versions of Python and Pyspark, similar to [here](https://github.com/fpgmaas/deptry/actions/runs/9938821113) with different OS and Python versions. Is it ok if we handle that in a separate PR? I will create an issue for this.

[the remark this answers is not in the document]
Comment (fpgmaas) [reply .github/workflows/ci.yml:26]: Created #109  and #110

File: pyproject.toml [hunk 3]
```diff
@@ -8,22 +8,22 @@ license = "MIT"
...
 [tool.poetry.group.mkdocs.dependencies]
-mkdocstrings-python = "^0.8.3"
-mkdocs-gen-files = "^0.4.0"
+mkdocstrings-python = "^1.7.3"
+mkdocs-gen-files = "^0.5.0"
 mkdocs-literate-nav = "^0.6.0"
 mkdocs-section-index = "^0.3.5"
 markdown-include = "^0.8.1"
-mkdocs = "^1.4.2"
+mkdocs = "^1.5.3"
```
[the remark this answers is not in the document]
Comment (fpgmaas) [reply pyproject.toml:23]: Sounds good to me, I will change it accordingly.

**Comment by fpgmaas** [comment]
> I left few comments

I think I resolved them!

**Review by SemyonSinchenko — APPROVED** [review]
LGTM, thank you!

**Pull request merged**